components/form-control/script.js

```diff
@@ -0,0 +1,313 @@
+const VALID_FORM_CONTROL_CUSTOM_ELEMENT_TAGS = new Set([
+    tags.GAMEFACE_CHECKBOX,
+    tags.GAMEFACE_DROPDOWN,
+    tags.GAMEFACE_RADIO_GROUP,
+    tags.GAMEFACE_RANGESLIDER,
+    tags.GAMEFACE_SWITCH
+]);
+
+class GamefaceFormControl extends HTMLElement {
+    constructor() {
+        super();
+        this.currentSubmitButton = null;
+    }
+
+    get method() {
+        return this.getAttribute('method') || formMethods.GET;
+    }
+
+    get action() {
+        return this.getAttribute('action') || '';
+    }
+
+    /**
+     * Will serialize the form data by traversing all the form element tree
+     * @returns {string}
+     */
+    getFormSerializedData() {
+        const params = new URLSearchParams();
+
+        //Serialize the data if there is any set on the submit button
+        this.serializeSubmitButtonElementData(params);
+        this.traverseFormElements(this, (element) => {
+            //We dont need data from invalid elements and elements that are having type "submit"
+            if (!this.isValidFromElement(element) || this.isElementUsedToSubmit(element)) return;
+
+            this.serializeElementData(element, params);
+        })
+
+        return params.toString();
+    }
+
+    /**
+     * Will serialize the data from the submit button if it has value attribute and it is <button> or <input type="submit"/>.
+     * @param {URLSearchParams} params
+     */
+    serializeSubmitButtonElementData(params) {
+        if (!this.currentSubmitButton) return;
+
+        const tagName = this.currentSubmitButton.tagName.toLowerCase();
+        if (tagName !== tags.BUTTON || tagName !== tags.INPUT) return;
+
+        if (this.currentSubmitButton.hasAttribute('name') && this.currentSubmitButton.hasAttribute('value')) {
+            params.append(this.currentSubmitButton.getAttribute('name'), this.currentSubmitButton.getAttribute('value'));
+        }
+    }
+
+    /**
+     * Will serialize data from a simple form element like input or textarea
+     * @param {HTMLElement} element
+     * @param {URLSearchParams} params
+     */
+    serializeSimpleElementData(element, params) {
+        if (element.hasAttribute('name') && element.value !== undefined) {
+            params.append(element.getAttribute('name'), element.value);
+        }
+    }
+
+    /**
+     * Will serialize the data from gameface-rangeslider component
+     * @param {HTMLElement} element - The gameface-rangeslider element
+     * @param {URLSearchParams} params
+     */
+    serializeRangeSliderData(element, params) {
+        if (element.hasAttribute('two-handles')) {
+            console.warn('gameface-rangeslider component does not support form data when "two-handles" attribute is used!');
+            return;
+        }
+
+        //The real rangeslider value is hidden inside _value[0]
+        if (element.hasAttribute('name') && element._value[0] !== undefined) {
+            params.append(element.getAttribute('name'), element._value[0]);
+        }
+    }
```

Comment:
Well you are right and no. I was thinking bout this and decided to make the data serialization in the form control because we have input, textarea and button elements that we can't add serialize methods to them. If we can't then the serialize methods will be spread between the form control and the components because serialize methods for the input, textarea and button will be included to the form control... Like this, as it is right now the serialize methods are in one place.

However, I can't see which approach is better.. maybe be should discuss this on the refinement meeting :D

Reply:
Yes, it is too bad that we can't add the validation to the native elements.
The downside of the current implementation is that we have component - specific logic that is outside of the component.
The ideal solution would be to have a serialize method that loops all form elements and gets the value from them like in Chromium: [the form creates the form data and passes it to the element](https://github.com/chromium/chromium/blob/c4d3c31083a2e1481253ff2d24298a1dfe19c754/third_party/blink/renderer/core/html/forms/html_form_element.cc#L573)
[and the element checks if its valid, if it has the required attributes and values and adds them to the form data](https://github.com/chromium/chromium/blob/c4d3c31083a2e1481253ff2d24298a1dfe19c754/third_party/blink/renderer/core/html/custom/element_internals.cc#L380)
This approach is very clean as the form does not implement element - specific logic, it has one method that serializes all elements as oppose to our current solution where we have methods for each form element type.
Most of our element specific methods contain validation, so I hope that once we move the validation the elements, we will be able to implement more abstract serialization where we simply call FormElement.getValue (with some exceptions for the native elements).